File: kline-indicator-client/src/types/indicator.ts
// 指标类型
export type IndicatorType = 'moshi_chanlun';

// 指标分类
export type IndicatorCategory = 'chanlun';

// 参数定义
export interface ParameterDef {
  name: string;
  type: 'int' | 'float' | 'bool' | 'string';
  required: boolean;
  default_value: number | string | boolean;
  min?: number;
  max?: number;
  description: string;
}

// 指标元信息
export interface IndicatorMetadata {
  type: IndicatorType;
  name: string;
  category: IndicatorCategory;
  description: string;
  params_def: ParameterDef[];
}

// 指标配置
export interface IndicatorConfig {
  type: IndicatorType;
  params: Record<string, number | string | boolean>;
}

// 指标值
export interface IndicatorValue {
  timestamp: string;
  value: number;
}

// 指标序列
export interface IndicatorSeries {
  name: string;
  values: IndicatorValue[];
}

// 分型类型
export type FractalType = 'top' | 'bottom' | 'TOP' | 'BOTTOM';

// 分型标记
export interface FractalMarker {
  index: number;
  timestamp: string;
  type: FractalType;
  price: number;
  zone?: [number, number];
}

// 笔方向
export type BiDirection = 'up' | 'down' | 'UP' | 'DOWN';

// 笔标记
export interface BiMarker {
  start_index: number;
  end_index: number;
  start_timestamp: string;
  end_timestamp: string;
  start_price: number;
  end_price: number;
  direction: BiDirection;
  length?: number;
  up_count?: number;    // 上涨K线数量
  down_count?: number;  // 下跌K线数量
  multiplier?: number;  // 级别倍数 0/1/2/4/8
  actual_retrace_time?: number; // 实际回调/反弹时间(分钟)
}

// 莫氏缠论标注点
export interface MoshiMarkPoint {
  type: 'L' | 'H';
  index: number;
  timestamp: string;
  price: number;
  level: string;
  multiplier: number;
}

// x1同级别走势
// pattern: "trend"(趋势), "convergent"(收敛型中枢), "divergent"(扩张型中枢)
export interface SameLevelTrend {
  type: 'up' | 'down';                          // 走势方向
  pattern: 'trend' | 'convergent' | 'divergent'; // 形态类型
  multiplier: number;                            // 走势所属级别: 1(x1), 2(x2), 4(x4), 8(x8)
  start_index: number;                          // 起始K线索引
  end_index: number;                            // 结束K线索引
  start_timestamp: string;                      // 起始时间
  end_timestamp: string;                        // 结束时间
  high_point: MoshiMarkPoint;                   // 走势最高点（上涨=最后H，下跌=第一H）
  low_point: MoshiMarkPoint;                    // 走势最低点（上涨=第一L，下跌=最后L）
  points: MoshiMarkPoint[];                     // 组成走势的所有L/H点序列
  // 级别升级相关字段
  upgraded?: boolean;                           // 是否已升级为父级别
  parent_points?: MoshiMarkPoint[];             // 升级后的父级别点序列
}

// 指标计算结果
export interface IndicatorResult {
  type: IndicatorType;
  name?: string;
  series?: IndicatorSeries[];
  fractal_markers?: FractalMarker[];
  bi_markers?: BiMarker[];
  extra?: {
    mark_points?: MoshiMarkPoint[];
    kline_type?: number;
    levels?: number[];
    same_level_trends?: SameLevelTrend[];  // x1同级别走势
  };
  computation_time_ms: number;
}

// 指标计算请求
export interface IndicatorCalculateRequest {
  market: number;
  code: string;
  klinetype: number;
  weight: number;
  count: number;
  indicators: IndicatorConfig[];
}

// 指标计算响应
export interface IndicatorCalculateResponse {
  code: number;
  message: string;
  stock_code: string;
  indicators: IndicatorResult[];
  computation_time_ms: number;
}

// 指标列表响应
export interface IndicatorListResponse {
  code: number;
  message: string;
  indicators: IndicatorMetadata[];
}

File: kline-indicator-service/internal/models/indicator.go
package models

// IndicatorType 指标类型
type IndicatorType string

const (
	IndicatorTypeMoshiChanlun   IndicatorType = "moshi_chanlun"   // 莫氏缠论
	IndicatorTypeChanlunFractal IndicatorType = "chanlun_fractal" // 缠论分型
	IndicatorTypeChanlunBi      IndicatorType = "chanlun_bi"      // 缠论笔
	IndicatorTypeMA             IndicatorType = "ma"              // 移动平均线
	IndicatorTypeEMA            IndicatorType = "ema"             // 指数移动平均线
	IndicatorTypeBOLL           IndicatorType = "boll"            // 布林带
)

// IndicatorCategory 指标分类
type IndicatorCategory string

const (
	CategoryChanlun IndicatorCategory = "chanlun" // 缠论指标
	CategoryBasic   IndicatorCategory = "basic"   // 基础指标
)

// ParameterDef 参数定义
type ParameterDef struct {
	Name         string      `json:"name"`
	Type         string      `json:"type"` // int, float, bool, string
	Required     bool        `json:"required"`
	DefaultValue interface{} `json:"default_value"`
	Min          interface{} `json:"min,omitempty"`
	Max          interface{} `json:"max,omitempty"`
	Description  string      `json:"description"`
}

// IndicatorMetadata 指标元信息
type IndicatorMetadata struct {
	Type        IndicatorType     `json:"type"`
	Name        string            `json:"name"`
	Category    IndicatorCategory `json:"category"`
	Description string            `json:"description"`
	ParamsDef   []ParameterDef    `json:"params_def"`
}

// IndicatorConfig 指标配置
type IndicatorConfig struct {
	Type   IndicatorType          `json:"type"`
	Params map[string]interface{} `json:"params"`
}

// IndicatorValue 指标数值（用于线形指标）
type IndicatorValue struct {
	Timestamp string  `json:"timestamp"`
	Value     float64 `json:"value"`
}

// IndicatorSeries 指标序列（用于多值指标如BOLL、MACD）
type IndicatorSeries struct {
	Name   string           `json:"name"`
	Values []IndicatorValue `json:"values"`
}

// FractalType 分型类型
type FractalType string

const (
	FractalTypeTop    FractalType = "TOP"
	FractalTypeBottom FractalType = "BOTTOM"
)

// FractalMarker 分型标记
type FractalMarker struct {
	Index     int         `json:"index"`
	Timestamp string      `json:"timestamp"`
	Type      string      `json:"type"`  // "top" or "bottom"
	Price     float64     `json:"price"`
	Zone      [2]float64  `json:"zone,omitempty"` // [下沿, 上沿]
}

// BiDirection 笔方向
type BiDirection string

const (
	BiDirectionUp   BiDirection = "UP"
	BiDirectionDown BiDirection = "DOWN"
)

// BiMarker 笔标记
type BiMarker struct {
	StartIndex        int     `json:"start_index"`
	EndIndex          int     `json:"end_index"`
	StartTimestamp    string  `json:"start_timestamp"`
	EndTimestamp      string  `json:"end_timestamp"`
	StartPrice        float64 `json:"start_price"`
	EndPrice          float64 `json:"end_price"`
	Direction         string  `json:"direction"`                    // "up" or "down"
	Length            int     `json:"length,omitempty"`             // 包含K线数
	UpCount           int     `json:"up_count,omitempty"`           // 上涨K线数量
	DownCount         int     `json:"down_count,omitempty"`         // 下跌K线数量
	Multiplier        int     `json:"multiplier"`                   // 级别倍数 0/1/2/4/8
	ActualRetraceTime float64 `json:"actual_retrace_time,omitempty"` // 实际回调/反弹时间(分钟)
}

// IndicatorResult 指标计算结果
type IndicatorResult struct {
	Type            IndicatorType          `json:"type"`
	Name            string                 `json:"name,omitempty"`
	Series          []IndicatorSeries      `json:"series,omitempty"`           // 线形数据
	FractalMarkers  []FractalMarker        `json:"fractal_markers,omitempty"`  // 分型标记
	BiMarkers       []BiMarker             `json:"bi_markers,omitempty"`       // 笔标记
	Extra           map[string]interface{} `json:"extra,omitempty"`            // 扩展数据
	ComputationTime int64                  `json:"computation_time_ms"`        // 计算耗时
}

// ParamDef 参数定义 (别名)
type ParamDef = ParameterDef

// IndicatorCalculateRequest 指标计算请求
type IndicatorCalculateRequest struct {
	Market     int               `json:"market"`
	Code       string            `json:"code"`
	KLineType  KLineType         `json:"klinetype"`
	Weight     WeightType        `json:"weight"`
	Count      int               `json:"count"`
	Indicators []IndicatorConfig `json:"indicators"`
}

// IndicatorCalculateResponse 指标计算响应
type IndicatorCalculateResponse struct {
	Code            int               `json:"code"`
	Message         string            `json:"message"`
	StockCode       string            `json:"stock_code"`
	Indicators      []IndicatorResult `json:"indicators"`
	ComputationTime int64             `json:"computation_time_ms"`
}
